File: 2024/three/main.py
"""
Advent of code day three
"""
import re

def mul_from_expr(expr: str) -> int:
    sum_prod = 0
    mul_pattern = r"mul\((\d+),(\d+)\)"
    matches = re.findall(mul_pattern, expr)
    for a, b in matches:
        sum_prod += int(a) * int(b)
    return sum_prod

def part_one(input: list[str]) -> int:
    sum_prod = 0
    for line in input:
        sum_prod += mul_from_expr(line)
    return sum_prod

class Tokenizer:
    def __init__(self):
        self.expr = ""
        self.tokens = ["do()"]

    def reset_or_add(self, last_char: str, char: str):
        if self.expr == "":
            return
        if self.expr[-1] != last_char:
            self.expr = ""
        else:
            self.expr += char

    def tokenize(self, input: str):
        for char in input:
            match char:
                case 'm' | 'd':
                    if self.expr != "":
                        self.expr = ""
                    else:
                        self.expr += char
                case 'u':
                    self.reset_or_add('m', char)
                case 'o':
                    self.reset_or_add('d', char)
                case 'n':
                    self.reset_or_add('o', char)
                case "'":
                    self.reset_or_add('n', char)
                case 'l':
                    self.reset_or_add('u', char)
                case 't':
                    self.reset_or_add("'", char)
                case '(':
                    if self.expr == "":
                        continue
                    if self.expr[-1] not in "otl":
                        self.expr = ""
                    else:
                        self.expr += char
                case ')':
                    if self.expr == "":
                        continue
                    if not self.expr[-1].isdigit() and self.expr[-1] != '(':
                        self.expr = ""
                    else:
                        self.expr += char
                        self.tokens.append(self.expr)
                        self.expr = ""
                case '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9':
                    if self.expr == "":
                        continue
                    if not self.expr[-1].isdigit() and self.expr[-1] not in "(,":
                        self.expr = ""
                    else:
                        self.expr += char
                case ',':
                    if self.expr == "":
                        continue
                    if not self.expr[-1].isdigit():
                        self.expr = ""
                    else:
                        self.expr += char
                case _:
                    if len(self.expr) != 0:
                        self.expr = ""

    def get_tokens(self) -> list[str]:
        return self.tokens

    def evaluate(self) -> int:
        do = False
        sum_prod = 0
        for token in self.tokens:
            if token == "do()":
                do = True
            elif token == "don't()":
                do = False
            elif do:
                sum_prod += mul_from_expr(token)
        return sum_prod

def part_two(input: list[str]) -> int:
    tokenizer = Tokenizer()
    for line in input:
        tokenizer.tokenize(line)
    return tokenizer.evaluate()

def main():
    with open("input", "r") as f:
        lines = f.readlines()   
    print(f"Part one answer: {part_one(lines)}")
    print(f"Part two answer: {part_two(lines)}")

if __name__=="__main__":
    main()

File: 2024/three/src/main.rs
//! Advent of code day three
#![allow(dead_code)]

use std::fs;
use regex::Regex;

fn mul_from_expr(expr: &str) -> i64 {
    let mut sum_prod = 0i64;
    let pattern = Regex::new(r"mul\((\d+),(\d+)\)").unwrap();
    for cap in pattern.captures_iter(expr) {
        let (_, vals): (&str, [&str; 2]) = cap.extract();
        let a: i64 = vals.iter().nth(0).unwrap().parse().unwrap();
        let b: i64 = vals.iter().nth(1).unwrap().parse().unwrap();
        sum_prod += a * b; 
    }
    return sum_prod;
}

fn part_one(input: &String) -> i64 {
    let mut sum_prod = 0i64;
    for line in input.lines() {
        sum_prod += mul_from_expr(&line);
    }
    return sum_prod;
}

struct Tokenizer {
    expr: String,
    tokens: Vec<String>
}

impl Tokenizer {
    pub fn new() -> Self {
        Tokenizer{ expr: String::new(), tokens: vec![String::from("do()")] }
    }

    fn reset_or_add(&mut self, last_char: char, chr: char) {
        if self.expr.is_empty() {
            return;
        }
        if !self.expr.ends_with(last_char) {
            self.expr.clear();
        } else {
            self.expr.push(chr);
        }
    }

    pub fn tokenize(&mut self, input: &str) {
        for chr in input.chars() {
            match chr {
                'm' | 'd' => {
                    if !self.expr.is_empty() {
                        self.expr.clear();
                    } else {
                        self.expr.push(chr);
                    }
                },
                'u' => self.reset_or_add('m', chr),
                'o' => self.reset_or_add('d', chr),
                'n' => self.reset_or_add('o', chr),
                '\'' => self.reset_or_add('n', chr),
                'l' => self.reset_or_add('u', chr),
                't' => self.reset_or_add('\'', chr),
                '(' => {
                    if self.expr.is_empty() {
                        continue;
                    }
                    if !"otl".contains(self.expr.chars().last().unwrap()) {
                        self.expr.clear();
                    } else {
                        self.expr.push(chr);
                    }
                },
                ')' => {
                    if self.expr.is_empty() {
                        continue;
                    }
                    let end_chr: char = self.expr.chars().last().unwrap();
                    if !end_chr.is_ascii_digit() && end_chr != '(' {
                        self.expr.clear();
                    } else {
                        self.expr.push(chr);
                        self.tokens.push(self.expr.clone());
                        self.expr = String::new();
                    }
                }
                '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' => {
                    if self.expr.is_empty() {
                        continue;
                    }
                    let end_chr = self.expr.chars().last().unwrap();
                    if !end_chr.is_ascii_digit() && !"(,".contains(end_chr) {
                        self.expr.clear();
                    } else {
                        self.expr.push(chr);
                    }
                }
                ',' => {
                    if self.expr.is_empty() {
                        continue;
                    }
                    if !self.expr.chars().last().unwrap().is_ascii_digit() {
                        self.expr.clear();
                    } else {
                        self.expr.push(chr);
                    }
                }
                _ => {
                    if !self.expr.is_empty() {
                        self.expr.clear();
                    }
                }
            }
        }
    }

    pub fn get_tokens(&self) -> Vec<String> {
        self.tokens.clone()
    }

    pub fn evaluate(&self) -> i64 {
        let mut do_cond = false;
        let mut sum_prod = 0i64;
        for token in self.tokens.iter() {
            if *token == "do()" {
                do_cond = true;
            } else if *token == "don't()" {
                do_cond = false;
            } else if do_cond {
                sum_prod += mul_from_expr(token);
            }
        }
        return sum_prod;
    }
}

fn part_two(input: &String) -> i64 {
    let mut tokenizer = Tokenizer::new();
    for line in input.lines() {
        tokenizer.tokenize(line);
    }
    tokenizer.evaluate()
}

fn main() {
    let input: String = fs::read_to_string("input").unwrap();
    // let input = String::from("mul(1,2)\nmul(3,4)\n");
    println!("Part one answer: {}", part_one(&input));
    println!("Part two answer: {}", part_two(&input));
}
